astropy/utils/decorators.py

```diff
@@ -863,7 +866,212 @@ def __get__(self, obj, objtype=None):
+
+
+@dataclasses.dataclass(frozen=True)
+class _dir_:
+    """``__dir__`` replacement that allows for filtering of entries.
+
+    Parameters
+    ----------
+    fget : callable[[object], Collection[str]] or None, optional
+        User-defined callable returning entries to remove to the enclosing
+        instance's ``__dir__``.
+    exclude : Collection[str], optional keyword-only
+        Entries to remove to the enclosing instance's ``__dir__``.
+
+    Examples
+    --------
+    The simplest usage of ``_dir_`` is to exclude select methods from an
+    instances ``__dir__``, for example if a `classmethod` only makes sense when
+    called from the class, and never from an instance.
+
+        >>> class Example:
+        ...    __dir__ = _dir_(exclude=['_private'])
+        ...    @classmethod
+        ...    def _private(cls):
+        ...        return
+        >>> ex = Example()
+        >>> "_private" in dir(ex)
+        False
+
+    The ``__dir__`` of the ``Example`` **class** is not overridden (that would
+    require messing with metaclasses), so ``_private`` is still visible on the
+    class:
+
+        >>> "_private" in dir(Example)
+        True
+
+    ``_dir_`` can also be used to decorate a user-defined ``__dir__`` override:
+
+        >>> class Example:
+        ...    @_dir_(exclude=['_private'])
+        ...    def __dir__(self):
+        ...        return super().__dir__() + ["special_method"]
+        ...    @classmethod
+        ...    def _private(cls):
+        ...        return
+        >>> ex = Example()
+        >>> "_private" in dir(ex)
+        False
+        >>> "special_method" in dir(ex)
+        True
+
+    Again, note that in Python ``__dir__`` for an instance does not override the
+    method for the class.
+
+        >>> "_private" in dir(Example)
+        True
+        >>> "special_method" in dir(Example)
+        False
+
+    Notes
+    -----
+    In Python 3.10+ ``exclude`` is actually keyword-only.
+
+    Python hides descriptors. To directly access an attribute on this
+    descriptor, e.g. ``exclude``, use
+    ``vars(<class>)["__dir__"].exclude``.
+    """
+
+    fget: Optional[Callable[[object], Collection[str]]] = dataclasses.field(default=None)
+    if sys.version_info >= (3, 10):
+        _: dataclasses.KW_ONLY
+    exclude: Set[str] = dataclasses.field(default_factory=set)
+
+    def __post_init__(self):
+        object.__setattr__(self, "__doc__", getattr(self.fget, "__doc__", self.__doc__))
+        object.__setattr__(self, "exclude", set(self.exclude))
+
+    def _overriden__dir__(_self, self: object) -> List[str]:
+        """`object.__dir__` with filtering of entries."""
+        members = set(object.__dir__(self) if _self.fget is None else _self.fget(self))
+        members -= _self.exclude
+        return sorted(members)
+
+    # -------------------------------------------
+    # Descriptor
+
+    def __get__(self, encl_inst: Optional[object], encl_cls: Optional[type]=None) -> MethodType:
```

Comment:
`None` is an `object`, so `Optional[object]` is the same type as `object`:
```suggestion
    def __get__(self, encl_inst: object, encl_cls: Optional[type]=None) -> MethodType:
```